src/DssSpell.t.sol

```diff
@@ -2105,309 +2134,746 @@ contract DssSpellTest is DSTest, DSMath {
+        assertEq(flipETHA.wards(address(cat))    , 0);
+        assertEq(flipETHA.wards(address(flipMom)), 0);
+
+        assertEq(vat.wards(address(clipETHA))    , 1);
+        assertEq(dog.wards(address(clipETHA))    , 1);
+        assertEq(clipETHA.wards(address(dog))    , 1);
+        assertEq(clipETHA.wards(address(end))    , 1);
+        assertEq(clipETHA.wards(address(clipMom)), 1);
+        assertEq(clipETHA.wards(address(esm)), 1);
+
+        assertEq(pipETH.bud(address(clipETHA)), 1);
+        assertEq(pipETH.bud(address(clipMom)), 1);
+
+        // Force max debt ceiling for ETH-A
+        hevm.store(
+            address(vat),
+            bytes32(uint256(keccak256(abi.encode(bytes32("ETH-A"), uint256(2)))) + 3),
+            bytes32(uint256(-1))
+        );
+
+        // Add balance to the test address
+        uint256 ilkAmt = 20 * WAD;
+
+        giveTokens(ETH, ilkAmt);
+        assertEq(ETH.balanceOf(address(this)), ilkAmt);
+
+        // Join to adapter
+        assertEq(vat.gem("ETH-A", address(this)), 0);
+        ETH.approve(address(joinETHA), ilkAmt);
+        joinETHA.join(address(this), ilkAmt);
+        assertEq(ETH.balanceOf(address(this)), 0);
+        assertEq(vat.gem("ETH-A", address(this)), ilkAmt);
+
+        // Generate new DAI to force a liquidation
+        (,uint256 rate, uint256 spot,,) = vat.ilks("ETH-A");
+        // dart max amount of DAI
+        int256 art = int256(mul(ilkAmt, spot) / rate);
+        vat.frob("ETH-A", address(this), address(this), address(this), int256(ilkAmt), art);
+        hevm.warp(block.timestamp + 1);
+        jug.drip("ETH-A");
+        assertEq(clipETHA.kicks(), 0);
+        dog.bark("ETH-A", address(this), address(this));
+        assertEq(clipETHA.kicks(), 1);
+
+        (,rate,,,) = vat.ilks("ETH-A");
+        uint256 debt = mul(mul(rate, uint256(art)), dog.chop("ETH-A")) / WAD;
+        hevm.store(
+            address(vat),
+            keccak256(abi.encode(address(this), uint256(5))),
+            bytes32(debt)
+        );
+        assertEq(vat.dai(address(this)), debt);
+        assertEq(vat.gem("ETH-A", address(this)), 0);
 
-        assertEq(vat.can(address(pauseProxy), address(daiJoin)), 1);
+        hevm.warp(block.timestamp + 20 minutes);
+        (, uint256 tab, uint256 lot, address usr,, uint256 top) = clipETHA.sales(1);
 
+        assertEq(usr, address(this));
+        assertEq(tab, debt);
+        assertEq(lot, ilkAmt);
+        assertTrue(mul(lot, top) > tab); // There is enough collateral to cover the debt at current price
+
+        vat.hope(address(clipETHA));
+        clipETHA.take(1, lot, top, address(this), bytes(""));
+
+        (, tab, lot, usr,,) = clipETHA.sales(1);
+        assertEq(usr, address(0));
+        assertEq(tab, 0);
+        assertEq(lot, 0);
+        assertEq(vat.dai(address(this)), 0);
+        assertEq(vat.gem("ETH-A", address(this)), ilkAmt); // What was purchased + returned back as it is the owner of the vault
+    }
```

Comment:
This should be converted into a generalized function. We have a bunch more of these to go through, so I think the effort is worthwhile. Feel free to leave as-is for this spell, but maybe for the next one.

Reply:
Yes, makes sense. Will see if I can do something for this week, but probably will happen for next one.